src/components/primitives/Text/Clamp.js

```diff
@@ -0,0 +1,149 @@
+import React, { PureComponent, Fragment } from "react";
+import { StyleSheet, Text, View } from "react-native";
+
+export default class Clamp extends PureComponent {
+  state = {
+    isTruncating: null,
+    containerWidth: null,
+    preEllipsisText: null,
+    restOfText: null,
+  };
+
+  handleSingleLineLayout = event => {
+    const singleLineWidth = event.nativeEvent.layout.width;
+    const {
+      numberOfLines,
+      children: text,
+      children: { length: totalCharCount },
+    } = this.props;
+    const { containerWidth } = this.state;
+
+    const averageCharsPerLine = getAverageCharsPerLine({
+      singleLineWidth,
+      totalCharCount,
+      containerWidth,
+    });
+
+    const { preEllipsisText, restOfText } = getText({
+      text,
+      numberOfLines,
+      averageCharsPerLine,
+    });
+
+    this.setState(state => ({
+      preEllipsisText,
+      restOfText,
+      isTruncating: false,
+    }));
+  };
+
+  handleContainerLayout = ({
+    nativeEvent: {
+      layout: { width },
+    },
+  }) => {
+    if (width === this.state.containerWidth) {
+      return;
+    }
+    this.setState(state => ({
+      containerWidth: width,
+      isTruncating: true,
+    }));
+  };
+
+  renderTextOnSingleLine = () => {
+    const { numberOfLines, style, ...rest } = this.props;
+    return (
+      <View style={styles.singleLineContainer}>
+        <Text
+          onLayout={this.handleSingleLineLayout}
+          numberOfLines={1}
+          style={[style, styles.singleLineText]}
+          {...rest}
+        >
+          {this.props.children}
+        </Text>
+      </View>
+    );
+  };
+
+  render() {
+    const { numberOfLines, ...rest } = this.props;
+    const { isTruncating, preEllipsisText, restOfText } = this.state;
+
+    return (
+      <View onLayout={this.handleContainerLayout} style={styles.container}>
+        {isTruncating ? (
+          this.renderTextOnSingleLine()
+        ) : (
+          <Fragment>
+            <PreEllipsisLines {...rest} text={preEllipsisText} />
+            <EllipsisLine {...rest} text={restOfText} />
+          </Fragment>
+        )}
+      </View>
+    );
+  }
+}
+
+const PreEllipsisLines = ({ text, ...rest }) => {
+  return <Text {...rest}>{text}</Text>;
+};
+
+const EllipsisLine = ({ text, ...rest }) => {
+  return text && text.length ? (
+    <Text numberOfLines={1} {...rest}>
+      {text}
+    </Text>
+  ) : null;
+};
+
+const getAverageCharsPerLine = ({
+  totalCharCount,
+  singleLineWidth,
+  containerWidth,
+}) => {
+  return (totalCharCount * containerWidth) / singleLineWidth;
+};
+
+const getLineBreakIndex = ({ charsPerLine, text }) => {
+  const draftSlice = text.slice(0, charsPerLine).lastIndexOf(" ");
+  return draftSlice;
+};
+
+const getText = config => {
+  const { text: fullText, numberOfLines, averageCharsPerLine } = config;
```

Comment:
I feel like this would be way simpler with a for loop. 
```
let preEllipsisText = [];
let restOfText = fullText;
for (let i = 0; i < numberOfLines; i++) {
  const lineBreakIndex = getLineBreakIndex({
    text: restOfText,
￼    charsPerLine,
￼  });

  preEllipsisText.push(restOfText.slice(0, lineBreakIndex));
  restOfText = restOfText.slice(lineBreakIndex);
}
```

You could actually avoid doing slices everytime by just iterating over getLineBreakIndex (assuming you don't actually need preEllipsisText to be an array):
```
let ellipsisIndex = 0;
for (let i = 0; i < numberOfLines; i++) {
   ellipsisIndex = getLineBreakIndex({ text: fullText, offset: ellipsisIndex, charsPerLine });
  if (ellipsisIndex == -1) {
    return { preEllipsisText: fulltext, restOfText: '' };
  }
}

return { preEllipsisText: fulltext.slice(0, ellipsisIndex), restOfText: fulltext.slice(ellipsisIndex);
}
```

The first one might be easier to read though.

Reply:
I agree that it's a little more cumbersome -- the virtue of the reducer approach vs for loop is stateless/point-free/immutable -- but it's possibly not worth it. Probably ways to make it clearer though  -- I can try to revisit